crates/biome_js_analyze/src/lint/nursery/use_import_extensions.rs

```diff
@@ -0,0 +1,222 @@
+    /// it should add based on the file extension of the current file and the import path.
+    /// When applying the suggested fix, make sure to verify that the file type is correct.
+    ///
+    pub UseImportExtensions {
+        version: "next",
+        name: "useImportExtensions",
+        recommended: false,
+        fix_kind: FixKind::Unsafe,
+    }
+}
+
+impl Rule for UseImportExtensions {
+    type Query = Ast<AnyJsImportSpecifierLike>;
+    type State = UseImportExtensionsState;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+
+        let file_ext = ctx.file_path().extension()?.to_str()?;
+
+        get_extensionless_import(file_ext, node)
+    }
+
+    fn diagnostic(_: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                state.module_name_token.text_range(),
+                markup! {
+                    "Add a file extension for relative imports."
+                },
+            )
+            .note(markup! {
+                "Explicit import improves compatibility with browsers and makes file resolution in tooling faster."
+            }),
+        )
+    }
+
+    fn action(ctx: &RuleContext<Self>, state: &Self::State) -> Option<JsRuleAction> {
+        let mut mutation = ctx.root().begin();
+
+        let (suggested_path, extension) = state.suggestion.clone()?;
+        let new_module_name = if ctx.as_preferred_quote().is_double() {
+            make::js_string_literal(&suggested_path)
+        } else {
+            make::js_string_literal_single_quotes(&suggested_path)
+        };
+
+        mutation.replace_element(
+            state.module_name_token.clone().into(),
+            new_module_name.into(),
+        );
+
+        Some(JsRuleAction {
+            category: ActionCategory::QuickFix,
+            applicability: Applicability::MaybeIncorrect,
+            message: markup! {
+                "Add potential import extension "<Emphasis>"."{extension}</Emphasis>"."
+            }
+            .to_owned(),
+            mutation,
+        })
+    }
+}
+
+pub struct UseImportExtensionsState {
+    suggestion: Option<(String, String)>,
+    module_name_token: SyntaxToken<JsLanguage>,
+}
+
+fn get_extensionless_import(
+    file_ext: &str,
+    node: &AnyJsImportSpecifierLike,
+) -> Option<UseImportExtensionsState> {
+    let module_name_token = node.module_name_token()?;
+    let module_path = inner_string_text(&module_name_token);
+    let path = Path::new(module_path.text());
+
+    if !matches!(
+        path.components().next(),
+        Some(Component::CurDir | Component::ParentDir)
+    ) || path.extension().is_some()
+    {
+        return None;
+    }
+
+    let has_query_or_hash = path
+        .components()
+        .last()
```

Comment:
We could call a single time `path.components()`
```suggestion
    let path_components = path.components();
    let first = path_components.next()?;
    if !matches!(
        first,
        Component::CurDir | Component::ParentDir
    ) || path.extension().is_some()
    {
        return None;
    }
    let last_component = path_components.last().unwrap_or(first);

    let has_query_or_hash = last_component
```